# Lab3 - Assignment Sentiment

Copyright: Vrije Universiteit Amsterdam, Faculty of Humanities, CLTL

This notebook describes the LAB-3 assignment of the Text Mining course. It is about sentiment analysis.

The aims of the assignment are:
* Learn how to run a rule-based sentiment analysis module (VADER)
* Learn how to run a machine learning sentiment analysis module (Scikit-Learn/ Naive Bayes)
* Learn how to run scikit-learn metrics for the quantitative evaluation
* Learn how to perform and interpret a quantitative evaluation of the outcomes of the tools (in terms of Precision, Recall, and F<sub>1</sub>)
* Learn how to evaluate the results qualitatively (by examining the data) 
* Get insight into differences between the two applied methods
* Get insight into the effects of using linguistic preprocessing
* Be able to describe differences between the two methods in terms of their results
* Get insight into issues when applying these methods across different  domains

In this assignment, you are going to create your own gold standard set from 50 tweets. You will the VADER and scikit-learn classifiers to these tweets and evaluate the results by using evaluation metrics and inspecting the data.

We recommend you go through the notebooks in the following order:
* **Read the assignment (see below)**
* **Lab3.2-Sentiment-analysis-with-VADER.ipynb**
* **Lab3.3-Sentiment-analysis.with-scikit-learn.ipynb**
* **Answer the questions of the assignment (see below) using the provided notebooks and submit**

In this assignment you are asked to perform both quantitative evaluations and error analyses:
* a quantitative evaluation concerns the scores (Precision, Recall, and F<sub>1</sub>) provided by scikit's classification_report. It includes the scores per category, as well as micro and macro averages. Discuss whether the scores are balanced or not between the different categories (positive, negative, neutral) and between precision and recall. Discuss the shortcomings (if any) of the classifier based on these scores
* an error analysis regarding the misclassifications of the classifier. It involves going through the texts and trying to understand what has gone wrong. It servers to get insight in what could be done to improve the performance of the classifier. Do you observe patterns in misclassifications?  Discuss why these errors are made and propose ways to solve them.

## Credits
The notebooks in this block have been originally created by [Marten Postma](https://martenpostma.github.io) and [Isa Maks](https://research.vu.nl/en/persons/e-maks). Adaptations were made by [Filip Ilievski](http://ilievski.nl).

## Part I: VADER assignments


### Preparation (nothing to submit):
To be able to answer the VADER questions you need to know how the tool works. 
* Read more about the VADER tool in [this blog](https://www.geeksforgeeks.org/python-sentiment-analysis-using-vader/).  
* VADER provides 4 scores (positive, negative, neutral, compound). Be sure to understand what they mean and how they are calculated.
* VADER uses rules to handle linguistic phenomena such as negation and intensification. Be sure to understand which rules are used, how they work, and why they are important.
* VADER makes use of a sentiment lexicon. Have a look at the lexicon. Be sure to understand which information can be found there (lemma?, wordform?, part-of-speech?, polarity value?, word meaning?) What do all scores mean? https://github.com/cjhutto/vaderSentiment/blob/master/vaderSentiment/vader_lexicon.txt) 


### [3.5 points] Question1:

Regard the following sentences and their output as given by VADER. Regard sentences 1 to 7, and explain the outcome **for each sentence**. Take into account both the rules applied by VADER and the lexicon that is used. You will find that some of the results are reasonable, but others are not. Explain what is going wrong or not when correct and incorrect results are produced. 

```
INPUT SENTENCE 1 I love apples
VADER OUTPUT {'neg': 0.0, 'neu': 0.192, 'pos': 0.808, 'compound': 0.6369}

INPUT SENTENCE 2 I don't love apples
VADER OUTPUT {'neg': 0.627, 'neu': 0.373, 'pos': 0.0, 'compound': -0.5216}

INPUT SENTENCE 3 I love apples :-)
VADER OUTPUT {'neg': 0.0, 'neu': 0.133, 'pos': 0.867, 'compound': 0.7579}

INPUT SENTENCE 4 These houses are ruins
VADER OUTPUT {'neg': 0.492, 'neu': 0.508, 'pos': 0.0, 'compound': -0.4404}

INPUT SENTENCE 5 These houses are certainly not considered ruins
VADER OUTPUT {'neg': 0.0, 'neu': 0.51, 'pos': 0.49, 'compound': 0.5867}

INPUT SENTENCE 6 He lies in the chair in the garden
VADER OUTPUT {'neg': 0.286, 'neu': 0.714, 'pos': 0.0, 'compound': -0.4215}

INPUT SENTENCE 7 This house is like any house
VADER OUTPUT {'neg': 0.0, 'neu': 0.667, 'pos': 0.333, 'compound': 0.3612}
```

### Sentence 1: "I love apples" — compound: 0.6369
The result is correct. "Love" has a high positive valence in the VADER lexicon, producing a positive compound score. No special rules are triggered and the output accurately reflects the sentiment.

### Sentence 2: "I don't love apples" — compound: -0.5216
The result is correct. VADER applies its negation rule, detecting "don't" and reversing the valence of "love". This correctly flips the sentiment to negative, showing that VADER handles simple negation well.

### Sentence 3: "I love apples :-)" — compound: 0.7579
The result is correct and stronger than Sentence 1. VADER recognizes ":-)" as a positive emoticon from its built-in emoticon lexicon and adds its valence on top of "love", boosting the compound score above that of Sentence 1.

### Sentence 4: "These houses are ruins" — compound: -0.4404
The result is incorrect. VADER assigns negative valence to "ruins" because the word is associated with destruction in its lexicon. However, here "ruins" is a noun referring to ancient remains, which is neutral. VADER cannot distinguish word senses or part-of-speech, leading to a wrong negative classification.

### Sentence 5: "These houses are certainly not considered ruins" — compound: 0.5867
The result is not necessarily right. The negation of "ruins" flips its valence to positive, and "certainly" acts as an intensifier that amplifies the score. However, the sentence is a neutral descriptive statement. The positive output is an artifact of rule interaction rather than genuine sentiment understanding.

### Sentence 6: "He lies in the chair in the garden" — compound: -0.4215
The result is incorrect. VADER assigns negative valence to "lies" because the word is associated with dishonesty in its lexicon. Here, however, "lies" means "to lie down", which is entirely neutral. This is a clear case of lexical ambiguity that VADER cannot resolve, as it has no mechanism to consider context or word sense.

### Sentence 7: "This house is like any house" — compound: 0.3612
The result is incorrect. VADER treats "like" as a positive sentiment word (as in "to like something") and assigns it positive valence. In this sentence "like" is used as a preposition meaning "similar to", carrying no sentiment. This again illustrates VADER's inability to handle lexical ambiguity.

### [Points: 2.5] Exercise 2: Collecting 50 tweets for evaluation
Collect 50 tweets. Try to find tweets that are interesting for sentiment analysis, e.g., very positive, neutral, and negative tweets. These could be your own tweets (typed in) or collected from the Twitter stream. If you have trouble accessing Twitter, try to find an existing dataset (on websites like kaggle or huggingface).

We will store the tweets in the file **my_tweets.json** (use a text editor to edit).
For each tweet, you should insert:
* sentiment analysis label: negative | neutral | positive (this you determine yourself, this is not done by a computer)
* the text of the tweet
* the Tweet-URL

from:
```
    "1": {
        "sentiment_label": "",
        "text_of_tweet": "",
        "tweet_url": "",
```
to:
```
"1": {
        "sentiment_label": "positive",
        "text_of_tweet": "All across America people chose to get involved, get engaged and stand up. Each of us can make a difference, and all of us ought to try. So go keep changing the world in 2018.",
        "tweet_url" : "https://twitter.com/BarackObama/status/946775615893655552",
    },
```

You can load your tweets with human annotation in the following way.

In [33]:
import json

In [44]:
import pandas as pd

# Load the CSV (adjust column names based on what you see above)
df = pd.read_csv(
    'data/twitter.csv',
    header=None,
    names=['id', 'entity', 'sentiment', 'tweet_text']
)

# Drop rows with missing tweet text
df = df.dropna(subset=['tweet_text'])


label_map = {
    'Positive': 'positive',
    'Negative': 'negative',
    'Neutral': 'neutral',
    'Irrelevant': 'neutral'  # map irrelevant to neutral
}
df['sentiment_label'] = df['sentiment'].map(label_map)

# try to get a balance: 17 positive, 17 negative, 16 neutral
pos = df[df['sentiment_label'] == 'positive'].sample(17, random_state=42)
neg = df[df['sentiment_label'] == 'negative'].sample(17, random_state=42)
neu = df[df['sentiment_label'] == 'neutral'].sample(16, random_state=42)

sampled = pd.concat([pos, neg, neu]).reset_index(drop=True)

my_tweets = {}
for i, row in sampled.iterrows():
    my_tweets[str(i + 1)] = {
        "sentiment_label": row['sentiment_label'],
        "text_of_tweet": row['tweet_text'], 
        "tweet_url": f"https://twitter.com/i/status/{row['id']}"
    }

# Save to my_tweets.json
with open('my_tweets.json', 'w', encoding='utf-8') as f:
    json.dump(my_tweets, f, indent=4, ensure_ascii=False)

print("Saved 50 tweets to my_tweets.json")

Saved 50 tweets to my_tweets.json


In [45]:
my_tweets = json.load(open('my_tweets.json'))

In [46]:
for id_, tweet_info in my_tweets.items():
    print(id_, tweet_info)
    break

1 {'sentiment_label': 'positive', 'text_of_tweet': 'My birthday consisted of crab legs, watching A Goofy Movie + Lady and the Tramp remake, and now I’m gonna play Overwatch with my best friend!! 🥳🥰', 'tweet_url': 'https://twitter.com/i/status/9309'}


### [5 points] Question 3:

Run VADER on your own tweets (see function **run_vader** from notebook **Lab2-Sentiment-analysis-using-VADER.ipynb**). You can use the code snippet below this explanation as a starting point. 
* [2.5 points] a. Perform a quantitative evaluation. Explain the different scores, and explain which scores are most relevant and why.
* [2.5 points] b. Perform an error analysis: select 10 positive, 10 negative and 10 neutral tweets that are not correctly classified and try to understand why. Refer to the VADER-rules and the VADER-lexicon. Of course, if there are less than 10 errors for a category, you only have to check those. For example, if there are only 5 errors for positive tweets, you just describe those.

In [47]:
def vader_output_to_label(vader_output):
    """
    map vader output e.g.,
    {'neg': 0.0, 'neu': 0.0, 'pos': 1.0, 'compound': 0.4215}
    to one of the following values:
    a) positive float -> 'positive'
    b) 0.0 -> 'neutral'
    c) negative float -> 'negative'
    
    :param dict vader_output: output dict from vader
    
    :rtype: str
    :return: 'negative' | 'neutral' | 'positive'
    """
    compound = vader_output['compound']
    
    if compound < 0:
        return 'negative'
    elif compound == 0.0:
        return 'neutral'
    elif compound > 0.0:
        return 'positive'
    
assert vader_output_to_label( {'neg': 0.0, 'neu': 0.0, 'pos': 1.0, 'compound': 0.0}) == 'neutral'
assert vader_output_to_label( {'neg': 0.0, 'neu': 0.0, 'pos': 1.0, 'compound': 0.01}) == 'positive'
assert vader_output_to_label( {'neg': 0.0, 'neu': 0.0, 'pos': 1.0, 'compound': -0.01}) == 'negative'

In [48]:
# Question 3a
from nltk.sentiment.vader import SentimentIntensityAnalyzer
from sklearn.metrics import classification_report

# Load tweets
with open('my_tweets.json') as f:
    my_tweets = json.load(f)

# Load VADER model
vader_model = SentimentIntensityAnalyzer()

tweets = []
all_vader_output = []
all_vader_labels = []
gold = []

# settings (to change for different experiments)
to_lemmatize = True 
pos = set()

for id_, tweet_info in my_tweets.items():
    the_tweet = tweet_info['text_of_tweet']
    vader_output = vader_model.polarity_scores(the_tweet) # run vader
    vader_label = vader_output_to_label(vader_output) # convert vader output to category
    
    tweets.append(the_tweet)
    all_vader_output.append(vader_output)
    all_vader_labels.append(vader_label)
    gold.append(tweet_info['sentiment_label'])
    
# use scikit-learn's classification report
print(classification_report(gold, all_vader_labels, digits=3))

              precision    recall  f1-score   support

    negative      0.389     0.412     0.400        17
     neutral      0.500     0.188     0.273        16
    positive      0.423     0.647     0.512        17

    accuracy                          0.420        50
   macro avg      0.437     0.415     0.395        50
weighted avg      0.436     0.420     0.397        50



### Question 3a quantitative evaluation

The classification report provides Precision, Recall, and F1-score per class, plus macro and weighted averages.

**Results interpretation:**

- Positive (F1 = 0.512): VADER performs best here. Positive words like excellent, awesome, love are strongly represented in the VADER lexicon and easy to detect. However, the recall score (0.647) significantly exceeds the precision score (0.423), meaning VADER over predicts positive, it flags many neutral and negative tweets as positive.
- Negative (F1 = 0.400): Moderate performance. VADER catches some negative tweets but misses many where the negativity is expressed through context like frustration and sarcasm rather than explicit negative words.
- Neutral (F1 = 0.273): Worst performance. The recall score is very low (0.188), indicating VADER rarely predicts "neutral", it tends to push tweets into positive or negative based on any sentiment word found in the text, even when the tweet as a whole is neutral.
- Macro average (F1 = 0.395): This averages the three class F1 scores equally, regardless of class size. Because our dataset is balanced (17/17/16), this is the most honest single summary of VADER's overall performance. Crucially, the neutral class's poor F1 (0.273) pulls the macro average down just as much as the positive class's relatively better F1 (0.512) pulls it up, nothing is hidden. The result, 0.395, is closer to the worst-performing class than the best, which tells you VADER's performance is uneven rather than consistently moderate.
- Weighted average (F1 = 0.397): This weights each class's F1 by how many examples it has before averaging. Because the dataset is nearly balanced, the weighted average comes out almost identical to the macro average (0.397 vs 0.395).  

The overall accuracy of 42%, which is only marginally above the random chance baseline of 33% for a 3-class problem, indicating VADER provides limited value on this dataset, particularly due to its poor performance on classifying neutral.

**Most Relevant Scores**

The macro F1 (0.395) and per-class F1 are most informative, since the dataset is roughly balanced across three classes. The macro F1 avoids hiding poor neutral performance behind strong positive performance. Accuracy alone is less useful because it equally weights a correct neutral prediction and a correct positive prediction.


In [49]:
# Question 3b
for category in ['positive', 'negative', 'neutral']:
    print(f"{category} — misclassified")
    for i, (tweet, pred, true, scores) in enumerate(zip(tweets, all_vader_labels, gold, all_vader_output)):
        if true == category and pred != true:
            print(f"ID {i+1} | Predicted: {pred} | True: {true} | Compound Score: {scores['compound']}")
            print(f"Tweet: {tweet}")
            print()

positive — misclassified
ID 4 | Predicted: negative | True: positive | Compound Score: -0.1901
Tweet: Procrastination with punk gets a bit ridiculous, but that means more time for Assassins Creed Valhalla.

ID 5 | Predicted: negative | True: positive | Compound Score: -0.1027
Tweet: Did you guys play Herdstein? because I'm obsessed & I feel like I'm cheating on Osrs

ID 6 | Predicted: neutral | True: positive | Compound Score: 0.0
Tweet: That’s right

ID 8 | Predicted: negative | True: positive | Compound Score: -0.7743
Tweet: @PlayStation PS I was with PS for 2 decades. ps3 and r. I will migrate to xbox series x (I already acquired it). was so sad and disappointed in ps5. still have not seen it in physical. and today's presentation October 4. untouchable console! its a joke! goodbye PS

ID 12 | Predicted: negative | True: positive | Compound Score: -0.4404
Tweet: amd gpu’s suck i will remember to switch over nvidia

ID 15 | Predicted: negative | True: positive | Compound Score: -0.186

### Question 3b Error Analysis

#### Positive tweets misclassified

**ID 4** — "Procrastination with punk gets a bit ridiculous, but that means more time for Assassins Creed Valhalla."  
Predicted **negative** (compound: –0.19). VADER picks up "ridiculous" (negative in its lexicon) and ignores the playful, enthusiastic context. The tweet is positively framed, but VADER does not know the context. The word "Ridiculous" here is treated as a straightforwardly negative word.

**ID 5** — "Did you guys play Herdstein? because I'm obsessed & I feel like I'm cheating on Osrs"  
Predicted **negative** (compound: –0.10). The word "cheating" has negative valence in the VADER lexicon. Here it is used figuratively to express enthusiasm, but VADER cannot resolve this figurative usage.

**ID 6** — "That's right"
Predicted **neutral** (compound: 0.0). This is too short for VADER to detect any sentiment, none of its words appear in the lexicon with non-zero valence. The positive label is inferred from conversational context that VADER has no access to.

**ID 8** — "@PlayStation PS I was with PS for 2 decades... was so sad and disappointed in ps5... goodbye PS"
Predicted **negative** (compound: –0.77). The human label is positive (maybe it is reflecting the positive purchase of Xbox), but the tweet contains many strongly negative words like sad and disappointed. The tweet reads negative even to humans without full context. VADER's negative prediction is arguably justified here.

**ID 12** — "amd gpu's suck i will remember to switch over nvidia"
Predicted **negative** (compound: –0.44). VADER correctly detects "suck" as negative. The human annotator labelled this positive, maybe beacuse it might be focusing on the switch to Nvidia, which is positive, but VADER reads the overall tone as negative, which is linguistically defensible.

**ID 15** — "2015 SHOCK OPS IS BATTLEFIELD AT IT'S BEST"
Predicted **negative** (compound: –0.19). The word "shock" has a slight negative valence in the VADER lexicon. The intended meaning is enthusiastic and positive. All-caps should boost intensity per VADER rules, but "shock" itself pulls the score down.

#### Negative tweets misclassified

**ID 20** — "@verizon. I cancelled ALL services. Please restore ALL of my accounts - as of the Right Now!. Like Now"  
Predicted **positive** (compound: +0.65). VADER scores "Right", "Now", and "Please" as positive and polite words. The aggressive, frustrated intent and repetition is not captured. VADER only sees the surface politeness instead of rage.

**ID 26** — "Amazing was how toxic the FIFA community is... Congrats @EASPORTSFIFA for creating one of the six most toxic communities"
Predicted **positive** (compound: +0.86). VADER sees "Amazing" and "Congrats" both strongly positive in its lexicon and misses that they are used sarcastically. VADER has no sarcasm detection. Ironic use of positive words cannot be captured.

**ID 28** — "Tiger tells @GOLFTV: "Stricks is addicted to Fortnite. I don't like Fortnite."" 
Predicted **positive** (compound: +0.36). The explicit negative statement "I don't like Fortnite" should be caught by VADER's negation rule. The word "addicted" kind of like a neutral-to-negative word create competing signals. VADER scores "like" positively and the negation rule is not strong enough to flip the whole compound score negative.

**ID 31** — "It keeps declining my transaction forms... I NEED HELP"  
Predicted **positive** (compound: +0.91). This is a customer complaint, but the tweet includes phrasing like "simply buy", "happy to regularly charge", and the word "help", all of which register as neutral or positive in the VADER lexicon. VADER entirely misses the frustration embedded in the sentence structure.

**ID 32** — "@Blizzard_Ent help me... I NEED TO MUST PLAY WARCRAFT III AT GOD FOR IT!!!!!!!!!!!!!!!"  
Predicted **positive** (compound: +0.82). All-caps and exclamation marks (VADER's intensification rules) amplify the presence of the word "PLAY" and the overall energy. VADER interprets high energy as positive sentiment, while the tweet is an expression of desperate frustration.

#### Neutral tweets misclassified

**ID 36** — "Just kill some things in Call of Duty, modern warfare with viewers"
Predicted **negative** (compound: –0.78). The word "kill" is strongly negative in the VADER lexicon. In a gaming context, "kill" is a technical term with no emotional valence. VADER lacks any domain adaptation for gaming vocabulary.

**ID 40** — "HIGH ELO Best Champions TIER List - League of Legends Patch 10.14"  
Predicted **positive** (compound: +0.82). The words "Best" and "High" are positive in the lexicon. Here they are used as category labels like "Best Champions" and "High Elo" with no evaluative sentiment intended by the author. VADER cannot distinguish descriptive usage from evaluative usage.

**ID 41** — "hell can't say this and it be shit my guy" 
Predicted **negative** (compound: –0.85). The words "hell" and "shit" are strongly negative in the VADER lexicon. This tweet uses internet slang in a neutral and playful way. VADER lacks register and slang awareness.

**ID 45** — "A ban for 2012 Battlefield 4 player Chiiroolaa has occurred"  
Predicted **negative** (compound: –0.87). The word "ban" receives a very negative score in the VADER lexicon. In this context it is a factual announcement. VADER treats any occurrence of "ban" as inherently negative rather than context-dependent.

**ID 49** — "New York charges Johnson & Johnson with insurance fraud over opioid claims"  
Predicted **negative** (compound: –0.71). Word like "fraud" is negative in the lexicon. This is a neutral news headline, just a factual reporting, not an expression of sentiment. VADER is not designed for news headlines, which often contain emotionally charged words in a purely informational context.

**Overall pattern:** 
1. Domain mismatch: words with clear domain-specific neutral meanings like "kill" and "ban" are assigned lexicon-level negative valence.
2. Sarcasm: positive words used sarcastically are not detected as ironic.
3. Context dependent tweets: tweets where sentiment relies on discourse context cannot be analysed by a lexicon-based system.

### [4 points] Question 4:
Run VADER on the set of airline tweets with the following settings:

* Run VADER (as it is) on the set of airline tweets 
* Run VADER on the set of airline tweets after having lemmatized the text
* Run VADER on the set of airline tweets with only adjectives
* Run VADER on the set of airline tweets with only adjectives and after having lemmatized the text
* Run VADER on the set of airline tweets with only nouns
* Run VADER on the set of airline tweets with only nouns and after having lemmatized the text
* Run VADER on the set of airline tweets with only verbs
* Run VADER on the set of airline tweets with only verbs and after having lemmatized the text

* [1 point] a. Generate for all separate experiments the classification report, i.e., Precision, Recall, and F<sub>1</sub> scores per category as well as micro and macro averages. **Use a different code cell (or multiple code cells) for each experiment.**
* [3 points] b. Compare the scores and explain what they tell you.
* - Does lemmatisation help? Explain why or why not.
* - Are all parts of speech equally important for sentiment analysis? Explain why or why not.

In [41]:
# Your code here


## Part II: scikit-learn assignments
### [4 points] Question 5
Train the scikit-learn classifier (Naive Bayes) using the airline tweets.

+ Train the model on the airline tweets with 80% training and 20% test set and default settings (TF-IDF representation, min_df=2)
+ Train with different settings:
    + with respect to vectorizing: TF-IDF ('airline_tfidf') vs. Bag of words representation ('airline_count') 
    + with respect to the frequency threshold (min_df). Carry out experiments with increasing values for document frequency (min_df = 2; min_df = 5; min_df =10) 
* [1 point] a. Generate a classification_report for all experiments
* [3 points] b. Look at the results of the experiments with the different settings and try to explain why they differ: 
    + which category performs best, is this the case for any setting?
    + does the frequency threshold affect the scores? Why or why not according to you?

In [ ]:
# Your code here


### [4 points] Question 6: Inspecting the best scoring features 

+ Train the scikit-learn classifier (Naive Bayes) model with the following settings (airline tweets 80% training and 20% test;  Bag of words representation ('airline_count'), min_df=2)
* [1 point] a. Generate the list of best scoring features per class (see function **important_features_per_class** below) [1 point]
* [3 points] b. Look at the lists and consider the following issues: 
    + [1 point] Which features did you expect for each separate class and why?
    + [1 point] Which features did you not expect and why ? 
    + [1 point] The list contains all kinds of words such as names of airlines, punctuation, numbers and content words (e.g., 'delay' and 'bad'). Which words would you remove or keep when trying to improve the model and why? 

In [ ]:
def important_features_per_class(vectorizer,classifier,n=80):
    class_labels = classifier.classes_
    feature_names =vectorizer.get_feature_names()
    topn_class1 = sorted(zip(classifier.feature_count_[0], feature_names),reverse=True)[:n]
    topn_class2 = sorted(zip(classifier.feature_count_[1], feature_names),reverse=True)[:n]
    topn_class3 = sorted(zip(classifier.feature_count_[2], feature_names),reverse=True)[:n]
    print("Important words in negative documents")
    for coef, feat in topn_class1:
        print(class_labels[0], coef, feat)
    print("-----------------------------------------")
    print("Important words in neutral documents")
    for coef, feat in topn_class2:
        print(class_labels[1], coef, feat) 
    print("-----------------------------------------")
    print("Important words in positive documents")
    for coef, feat in topn_class3:
        print(class_labels[2], coef, feat) 

# example of how to call from notebook:
#important_features_per_class(airline_vec, clf)

### [Optional! (will not  be graded)] Question 7
Train the model on airline tweets and test it on your own set of tweets
+ Train the model with the following settings (airline tweets 80% training and 20% test;  Bag of words representation ('airline_count'), min_df=2)
+ Apply the model on your own set of tweets and generate the classification report
* [1 point] a. Carry out a quantitative analysis.
* [1 point] b. Carry out an error analysis on 10 correctly and 10 incorrectly classified tweets and discuss them
* [2 points] c. Compare the results (cf. classification report) with the results obtained by VADER on the same tweets and discuss the differences.

### [Optional! (will not be graded)] Question 8: trying to improve the model
* [2 points] a. Think of some ways to improve the scikit-learn Naive Bayes model by playing with the settings or applying linguistic preprocessing (e.g., by filtering on part-of-speech, or removing punctuation). Do not change the classifier but continue using the Naive Bayes classifier. Explain what the effects might be of these other settings 
+ [1 point] b. Apply the model with at least one new setting (train on the airline tweets using 80% training, 20% test) and generate the scores
* [1 point] c. Discuss whether the model achieved what you expected.

## End of this notebook